### ISSUE TITLE
Dependencies on $CC are flaky

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2877]
This issue was originally created at: 2012-10-10 07:42:56.
This issue was reported by: `marcnarc`.
marcnarc said at 2012-10-10 07:42:56
>SCons doesn't always pick up a dependency on $CC.  Here's a simple SConstruct:
> 
```python
e = DefaultEnvironment()
mycc = e.Command(
    target="mycc",
    source="/usr/bin/gcc",
    action="cp $SOURCE $TARGET",
)
e.Replace(CC=mycc[0].abspath)
e.Program(target="testapp", source="test.c")
```
> 
>The above works fine:
```console
$ scons --tree=prune
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
cp /usr/bin/gcc mycc
/usr/home/marcnarc/Code/foo/sconsdep/mycc -o test.o -c test.c
/usr/home/marcnarc/Code/foo/sconsdep/mycc -o testapp test.o
+-.
  +-SConstruct
  +-mycc
  | +-/usr/bin/gcc
  | +-/bin/cp
  +-test.c
  +-test.o
  | +-test.c
  | +-[mycc]
  +-testapp
    +-[test.o]
    +-[mycc]
scons: done building targets.
```
 
>However, it breaks with seemingly innocent changes:
> 
> * Put mycc inside a tools/ directory (`target="tools/mycc"`).
> * Change the program's name from "testapp".  Some names work but others don't. Here are some names that don't work: aragon frodo gimli legolas boromir. And here are some names that do work: waragon, phrodo, qgimli plegolas sauron.
> * Run scons with the -q option.
> 
>What breaks is the dependency on `mycc`.  In each failing case the tree shows no dependency from `test.o` on `mycc`:
```console
...
+-.
  +-SConstruct
  +-test.c
  +-test.o
  | +-test.c
  +-testapp
  | +-[test.o]
  +-tools
    +-tools/mycc
      +-/usr/bin/gcc
      +-/bin/cp
```

>(I also see this problem with SCons 1.2.0.)

marcnarc said at 2012-10-10 12:28:32
>Created an attachment (id=905)
>taskmaster trace from a successful run

marcnarc said at 2012-10-10 12:28:55
>Created an attachment (id=906)
>taskmaster trace from a failed run

marcnarc said at 2012-10-10 12:37:50
>Created an attachment (id=907)
>taskmaster trace when the program's name is "gimli"

marcnarc said at 2012-10-10 12:38:48
>Attachment taskmaster.fail.trace is from a run where the mycc target is "tools/mycc".
> 
>Attachment taskmaster.fail-gimli.trace is from a run where the mycc target is just "mycc" but the program's target is "gimli".

marcnarc said at 2012-10-22 07:24:28
>Johan Holmberg noted on scons-users that all the working target names are lexically > than "mycc".
> 
>I've verified that the lexicographic order of the target names is significant: Renaming the mycc target to "Amycc" fixes the broken program names ("aragon" et al.).

marcnarc said at 2012-10-22 07:28:17
>A bit more info:
> 
>Specifying the target name (e.g. "testapp") on the scons command line always produces a failure, regardless of the name of the mycc target.

Votes for this issue: 1.

marcnarc attached [taskmaster.good.trace](https://github.com/SCons/tigris-issue-attachments/blob/master/2877/905/taskmaster.good.trace) at 2012-10-10 12:28:32.
>taskmaster trace from a successful run

marcnarc attached [taskmaster.fail.trace](https://github.com/SCons/tigris-issue-attachments/blob/master/2877/906/taskmaster.fail.trace) at 2012-10-10 12:28:55.
>taskmaster trace from a failed run

marcnarc attached [taskmaster.fail-gimli.trace](https://github.com/SCons/tigris-issue-attachments/blob/master/2877/907/taskmaster.fail-gimli.trace) at 2012-10-10 12:37:50.
>taskmaster trace when the program's name is "gimli"

**Comment by mwichmann** [comment]
This is just weird, why should there be a lexicograhical relationship to the pathname of a tool?  Anyway, I try to reproduce it, it doesn't work like this, now if doesn't "make" the tool at all.

**Comment by bdbaddog** [comment]
It seems you can't just cp gcc to another name in another location anymore.. 

```
$ python ~/devel/scons//git/as_scons/scripts/scons.py
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
cp /usr/bin/gcc mycc
/home/bdbaddog/devel/scons/git/scons-bugswork/2877/mycc -o test.o -c test.c
mycc: fatal error: cannot execute 'cc1': execvp: No such file or directory
compilation terminated.
scons: *** [test.o] Error 1
scons: building terminated because of errors.
```

**Comment by bdbaddog** [comment]
ugh. I can repro.
You can see the files here: https://github.com/bdbaddog/scons-bugswork/tree/main/2877

This is via the copy to tools dir referred to above.

**Comment by mwichmann** [comment]
Yeah, I was seeing the gcc problem as well, probably for demo purposes should pick a different tool.  Also if the action uses `Copy` it should figure out whether mkdir needed or not without having to wire it in? So maybe this isn't actually a problem? Using Copy it now looks like this - I see the tool in the deps where it's expected. 
```console
$ scons --tree=prune
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Copy("bin/mycc", "/usr/bin/gcc")
/home/mats/github/scons/exp/issue2877/bin/mycc -o test.o -c test.c
mycc: fatal error: cannot execute 'cc1': execvp: No such file or directory
compilation terminated.
scons: *** [test.o] Error 1
+-.
  +-SConstruct
  +-bin
  | +-bin/mycc
  |   +-/usr/bin/gcc
  +-test.c
  +-test.o
  | +-test.c
  | +-[bin/mycc]
  +-testapp
    +-[test.o]
    +-[bin/mycc]
scons: building terminated because of errors.
```

**Comment by bdbaddog** [comment]
@mwichmann  try it with my sample code linked above. I still see the issue.
I think it's an issue with the implicit command dependencies...

**Comment by mwichmann** [comment]
You're right, the lex-order-matters thing is still there, I didn't even notice in my test (pasted above) that I dodged it by using `bin/toolname` instead of `tools/toolname`. Just like the original report, I can toggle success/fail by swapping the subdirectory names between bin and tools...  weird.

**Comment by mwichmann** [comment]
SCons must be sorting something in order to get reproducibility, and thus somehow introducing this issue?
